marshal_test.go

```diff
@@ -0,0 +1,379 @@
+		}{
+			F1: newString("some random user"),
+			F2: newString("42"),
+		}
+		return &t
+	},
+	expectURLString: "http://localhost:8081/some%20random%20user?age=42",
+}, {
+	about:     "MarshalText called on TextMarhsalers",
+	urlString: "http://localhost:8081/:param1/:param2",
+	val: func() interface{} {
+		v := struct {
+			F1 testMarshaler  `httprequest:"param1,path"`
+			F2 *testMarshaler `httprequest:"param2,path"`
+			F3 testMarshaler  `httprequest:"param3,form"`
+			F4 *testMarshaler `httprequest:"param4,form"`
+		}{
+			F1: "test1",
+			F2: (*testMarshaler)(newString("test2")),
+			F3: "test3",
+			F4: (*testMarshaler)(newString("test4")),
+		}
+		return &v
+	},
+	expectURLString: "http://localhost:8081/test_test1/test_test2?param3=test_test3&param4=test_test4",
+}, {
+	about:     "MarshalText called on TextMarhsalers",
+	urlString: "http://localhost:8081/user/:name:surname",
+	val: func() interface{} {
+		v := struct {
+			F1 testMarshaler  `httprequest:"name,path"`
+			F2 *testMarshaler `httprequest:"surname,path"`
+		}{
+			F1: "name",
+			F2: (*testMarshaler)(newString("surname")),
+		}
+		return &v
+	},
+	expectURLString: "http://localhost:8081/user/test_nametest_surname",
+}, {
+	about:     "MarshalText not called on values that do not implement TextMarshaler",
+	urlString: "http://localhost:8081/user/:name:surname",
+	val: func() interface{} {
+		v := struct {
+			F1 notTextMarshaler  `httprequest:"name,path"`
+			F2 *notTextMarshaler `httprequest:"surname,path"`
+		}{
+			F1: "name",
+			F2: (*notTextMarshaler)(newString("surname")),
+		}
+		return &v
+	},
+	expectURLString: "http://localhost:8081/user/namesurname",
+}, {
+	about:     "MarshalText returns an error",
+	urlString: "http://localhost:8081/user/:name:surname",
+	val: func() interface{} {
+		v := struct {
+			F1 testMarshaler  `httprequest:"name,path"`
+			F2 *testMarshaler `httprequest:"surname,path"`
+		}{
+			F1: "",
+			F2: (*testMarshaler)(newString("surname")),
+		}
+		return &v
+	},
+	expectError: "cannot marshal field: empty string",
+}, {
+	about:     "[]string field form value",
+	urlString: "http://localhost:8081/user",
+	val: func() interface{} {
+		v := struct {
+			F1 []string `httprequest:"users,form"`
+		}{
+			F1: []string{"user1", "user2", "user3"},
+		}
+		return &v
+	},
+	expectURLString: "http://localhost:8081/user?users=user1&users=user2&users=user3",
+}, {
+	about:     "nil []string field form value",
+	urlString: "http://localhost:8081/user",
+	val: func() interface{} {
+		v := struct {
+			F1 *[]string `httprequest:"users,form"`
+		}{
+			F1: nil,
+		}
+		return &v
+	},
+	expectURLString: "http://localhost:8081/user",
+}, {
+	about:     "[]string field fails to marshal to path",
+	urlString: "http://localhost:8081/user/:users",
+	val: func() interface{} {
+		v := struct {
+			F1 []string `httprequest:"users,path"`
+		}{
+			F1: []string{"user1", "user2", "user3"},
+		}
+		return &v
+	},
+	expectError: ".*invalid target type.*",
```

Comment:
Let's specify the whole error message here - it's good to see whether the error looks appropriate or not.